include/range/v3/view/iota.hpp

```diff
@@ -44,399 +43,359 @@ namespace ranges
-        {}
     };
 
-    template<typename From>
-    struct iota_view<From, void>
-      : view_facade<iota_view<From, void>, infinite>
+#if RANGES_CXX_DEDUCTION_GUIDES >= RANGES_CXX_DEDUCTION_GUIDES_17
+    CPP_template(typename From, typename To)(
+        requires WeaklyIncrementable<From> && Semiregular<To> &&
+            (!Integral<From> || !Integral<To> ||
+                std::is_signed<From>::value == std::is_signed<To>::value))
+    closed_iota_view(From, To) -> closed_iota_view<From, To>;
+#endif
+
+    template<typename From, typename To /* = unreachable_sentinel_t*/>
+    struct RANGES_EMPTY_BASES iota_view
+      : view_facade<
+            iota_view<From, To>,
+            Same<To, unreachable_sentinel_t> ? infinite : finite>
+      , private iota_view_detail::adl_hook
     {
     private:
-        using incrementable_tag_of = ranges::incrementable_tag_of<From>;
         friend range_access;
-        using difference_type_ = detail::iota_difference_t<From>;
-
-        From value_;
+        From from_ = From();
+        RANGES_NO_UNIQUE_ADDRESS To to_ = To();
 
-        From read() const
+        struct cursor;
+        struct sentinel
         {
-            return value_;
-        }
-        void next()
-        {
-            ++value_;
-        }
-        constexpr bool equal(default_sentinel_t) const
+        private:
+            friend struct cursor;
+            RANGES_NO_UNIQUE_ADDRESS To to_;
+        public:
+            sentinel() = default;
+            constexpr explicit sentinel(To to)
+              : to_(std::move(to))
+            {}
+        };
+
+        struct cursor
         {
-            return false;
-        }
-        CPP_member
-        auto equal(iota_view const &that) const -> CPP_ret(bool)(
-            requires Incrementable<From>)
+        private:
+            friend range_access;
+            From from_;
+
+            From read() const
+            {
+                return from_;
+            }
+            void next()
+            {
+                ++from_;
+            }
+            bool equal(sentinel const &that) const
+            {
+                return from_ == that.to_;
+            }
+            CPP_member
+            auto equal(cursor const &that) const ->
+                CPP_ret(bool)(
+                    requires EqualityComparable<From>)
+            {
+                return that.from_ == from_;
+            }
+            CPP_member
+            auto prev() -> CPP_ret(void)(
+                requires Decrementable_<From>)
+            {
+                --from_;
+            }
+            CPP_member
+            auto advance(iter_difference_t<From> n) -> CPP_ret(void)(
+                requires Advanceable_<From>)
+            {
+                from_ += n;
+            }
+            CPP_member
+            auto distance_to(cursor const &that) const ->
+                CPP_ret(iter_difference_t<From>)(
+                    requires Advanceable_<From>)
```

Comment:
Not strictly to spec: the relational operators of the resulting `basic_iterator` effectively require *`Advanceable`*, - since they're implemented in terms of `distance_to` - whereas in the working draft those relational operators only require `StrictTotallyOrdered<From>`.